File: exercise/C#/day02/Games/FizzBuzz.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using LanguageExt;

namespace Games
{
    public class FizzBuzz
    {
        private readonly IDictionary<int, string> _mapping;

        public FizzBuzz(IDictionary<int, string> mapping)
        {
            _mapping = mapping
                .OrderBy(x => x.Key)
                .ToDictionary(x => x.Key, x => x.Value);
        }

        public const int Min = 1;
        public const int Max = 100;


        private static string ConvertSafely(int input, IDictionary<int, string> mapping)
        {
            var sb = mapping
                .Where(p => Is(p.Key, input))
                .Map(kvp => kvp.Value)
                .Aggregate(new StringBuilder(), (sb, s) => sb.Append(s));
            return sb.Length > 0 ? sb.ToString() : input.ToString();
        }

        private static bool Is(int divisor, int input) => input % divisor == 0;

        private static bool IsOutOfRange(int input) => input is < Min or > Max;

        public Option<string> Convert(int input)
            => IsOutOfRange(input)
                ? Option<string>.None
                : ConvertSafely(input, _mapping);
    }
}

File: exercise/java/day02/src/main/java/games/FizzBuzz.java
package games;

import io.vavr.collection.Map;
import io.vavr.control.Option;

import static io.vavr.control.Option.none;
import static io.vavr.control.Option.some;

public class FizzBuzz {
    public static final int MIN = 1;
    public static final int MAX = 100;
    private final Map<Integer, String> mapping;

    public FizzBuzz(Map<Integer, String> mapping) {
        this.mapping = mapping;
    }

    public Option<String> convert(int input) {
        return isOutOfRange(input)
                ? none()
                : some(convertSafely(input));
    }

    private String convertSafely(Integer input) {
        return mapping
                .find(p -> is(p._1, input))
                .map(p -> p._2)
                .getOrElse(input.toString());
    }

    private static boolean is(Integer divisor, Integer input) {
        return input % divisor == 0;
    }

    private static boolean isOutOfRange(Integer input) {
        return input < MIN || input > MAX;
    }
}